### PR TITLE
Treat package as resilient as non-frozen public by default

### PR DESCRIPTION
**Pull request by elsh**
Currently `package` types are treated non-resilient by default. This PR changes that to treat them as 
resilient as non-frozen `public` by default. If the defining module is built resiliently, treat the package 
decl as resilient, just as we do for `public` decls. Access methods to `package` should be the same 
as for non-frozen `public`, i.e. indirect. 
Added tests for SIL/IR wrt serialization and indirect access with/out resilience enabled.

Resolves rdar://118947451

### CODE REVIEW

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -4915,7 +4915,7 @@ bool NominalTypeDecl::isFormallyResilient() const {
   // Private, (unversioned) internal, and package types always have a
```
Comment (CodaFi) [inline lib/AST/Decl.cpp:4915]: Nit: This comment should strike `package` here.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

File: test/IRGen/package_resilience.swift [hunk 2]
```diff
@@ -214,29 +226,43 @@ package class ClassWithResilientThenEmpty {
...
   return MemoryLayout<Size>.size
 }
 
 // CHECK-LABEL: define{{.*}} swiftcc {{i32|i64}} @"$s18package_resilience40memoryLayoutDotStrideWithResilientStructSiyF"()
 public func memoryLayoutDotStrideWithResilientStruct() -> Int {
-  // CHECK: ret [[INT]] [[#WORDSIZE + WORDSIZE]]
+// CHECK: [[FIELD_VALUE:%.*]] = load ptr, ptr {{.*}},
+// CHECK-NEXT: [[FIELD_PTR:%.*]] = getelementptr inbounds %swift.vwtable, ptr [[FIELD_VALUE]],
+// CHECK-NEXT: [[SIZE:%.*]] = load [[INT]], ptr [[FIELD_PTR]],
+// CHECK-NEXT: ret [[INT]] [[SIZE]]
```
Comment (CodaFi) [inline test/IRGen/package_resilience.swift:241]: These paths have regressed

Comment (elsh) [reply test/IRGen/package_resilience.swift:241]: If I return MemoryLayout<PublicSize>.size, where PublicSize is defined as:
```
public struct PublicSize {
  public var w: Int
  public let h: Int 

  public init(w: Int, h: Int) {
    self.w = w
    self.h = h
  }

  public func method() {}
  public mutating func mutantMethod() {}
}
```
I get the following result with and without this PR.
```
define swiftcc i64 @"$s18package_resilience46memoryLayoutDotStrideWithPublicResilientStructSiyF"() #0 {
entry:
  %0 = call swiftcc %swift.metadata_response @"$s16resilient_struct10PublicSizeVMa"(i64 0) #14
  %1 = extractvalue %swift.metadata_response %0, 0
  %2 = getelementptr inbounds ptr, ptr %1, i64 -1
  %.valueWitnesses = load ptr, ptr %2, align 8, !invariant.load !23, !dereferenceable !24
  %3 = getelementptr inbounds %swift.vwtable, ptr %.valueWitnesses, i32 0, i32 8
  %size = load i64, ptr %3, align 8, !invariant.load !23
  ret i64 %size
}
```

This PR is to make the `package` just as resilient as non-frozen `public`. Optimization PR to make it non-resilient will follow later.

File: test/IRGen/package_resilience.swift [hunk 3]
```diff
@@ -214,29 +226,43 @@ package class ClassWithResilientThenEmpty {
...
   return MemoryLayout<Size>.size
```
Comment (xymus) [inline test/IRGen/package_resilience.swift:242]: It looks like there was a copy paste mishap here in the original code. I would expect this test to call `.stride`. As is the test is identical to the previous use case.

File: test/Sema/package_resilience_access.swift [hunk 4]
```diff
@@ -63,141 +286,263 @@ package func f(_ arg: PublicStruct) -> Int {
...
-// CHECK-NEXT:   debug_value %1 : $*PkgStructGeneric<T>, let, name "arg", argno 1, expr op_deref // id: %2
-// CHECK-NEXT:   %3 = struct_element_addr %1 : $*PkgStructGeneric<T>, #PkgStructGeneric.data // user: %4
-// CHECK-NEXT:   copy_addr %3 to [init] %0 : $*T                 // id: %4
-// CHECK-NEXT:   %5 = tuple ()                                   // user: %6
-// CHECK-NEXT:   return %5 : $()                                 // id: %6
-// CHECK-NEXT: } // end sil function '$s6Client1myx5Utils16PkgStructGenericVyxGlF'
+// CLIENT-RES-LABEL: // m<A>(_:)
+// CLIENT-RES-NEXT: sil [ossa] @$s6Client1myx5Utils16PkgStructGenericVyxGlF : $@convention(thin) <T> (@in_guaranteed PkgStructGeneric<T>) -> @out T {
+// CLIENT-RES-NEXT: // %0 "$return_value"                             // user: %6
+// CLIENT-RES-NEXT: // %1 "arg"                                       // users: %4, %2
```
Comment (xymus) [inline test/Sema/package_resilience_access.swift:365]: Are these `// users:` comment necessary to the test? John recently warned against superfluously strict SIL tests and pointed out most of these can usually be removed if they're not what's actually being tested.

Comment (elsh) [reply test/Sema/package_resilience_access.swift:365]: Not necessary. I'll remove these.

File: test/IRGen/package_resilience.swift [hunk 5]
```diff
@@ -10,24 +10,35 @@
...
+
 // RUN: %target-swift-frontend -package-name MyPkg -I %t -emit-ir -enable-library-evolution -O %t/package_resilience.swift -package-name MyPkg
 // REQUIRES: objc_codegen
 // REQUIRES: OS=macosx || OS=ios || OS=tvos || OS=watchos
 // REQUIRES: CPU=x86_64 || CPU=arm64
 
-// CHECK: @"$s18package_resilience26ClassWithResilientPropertyC1p16resilient_struct5PointVvpWvd" = constant [[INT]] {{8|16}}, align [[#WORDSIZE]]
-// CHECK: @"$s18package_resilience26ClassWithResilientPropertyC1s16resilient_struct4SizeVvpWvd" = constant [[INT]] {{32|16}}, align [[#WORDSIZE]]
+// CHECK: @"$s18package_resilience26ClassWithResilientPropertyC1p16resilient_struct5PointVvpWvd" = hidden global [[INT]] 0,
+// CHECK: @"$s18package_resilience26ClassWithResilientPropertyC1s16resilient_struct4SizeVvpWvd" = hidden global [[INT]] 0,
```
Comment (elsh) [inline test/IRGen/package_resilience.swift:26]: FYI I get `hidden global` for a public decl as well.

For example, 
```
public struct PublicSize {
  public var w: Int // should have getter and setter
  public let h: Int // getter only

  public init(w: Int, h: Int) {
    self.w = w
    self.h = h
  }

  public func method() {}
  public mutating func mutantMethod() {}
}

```

referenced in 

```
public class PublicClassWithPbProperty {
  public let p: PublicPoint
  public let s: PublicSize
  public let color: Int32

  public init(p: PublicPoint, s: PublicSize, color: Int32) {
    self.p = p
    self.s = s
    self.color = color
  }
}

```
results in 
```
@"$s18package_resilience25PublicClassWithPbPropertyC1p16resilient_struct0C5PointVvpWvd" = hidden global i64 0, align 8
```

File: test/IRGen/package_resilience.swift (hunk 3 above)
Comment (CodaFi) [reply test/IRGen/package_resilience.swift:242]: Yes, this should be stride.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test linux

**Review by xymus — APPROVED** [review]

**Pull request merged**